Update verification message logic

dmcgowan opened the pull request:
Only show the verification message if all the tarsum checks pass and the image manifest is verified. No longer return an error when a tarsum verification fails, just reset the verification flag. Tarsum verification is less meaningful without a verified manifest and therefore it should not cause an error.

This will make the last verification more meaningful since it will only be shown if both the manifest and layer checksums are verified.

tianon commented:
+1! :heart:

dmp42 commented:
LGTM

dmp42 commented:
cc @icecrime @jfrazelle @vbatts

jlhawn commented:
Once the tests are updated L<sup>G<sup>T<sup>M</sup></sup></sup>

dmcgowan commented:
May need to skip the test since the test is not pulling a verified image (signed by unknown key). I will investigate adding the key to the keychain, we currently don't have that flow defined for tests.

Going to make another update to also only display the bottom message if layers downloaded.  If no layers were downloaded the message is also meaningless, since nothing was verified.

dmcgowan commented:
Updated the test to pull from the central repository which should be a fully verified image. Pulling from a private registry will not yet unless we manually update the keychain, can probably add such a test in the future. Using the central repository will also be a good test if a key expiration occurs and images cannot be verified.

jfrazelle commented:
LGTM (I do not mind that we are now pulling from a central registry) hello-world is pretty small but idk what the others think because it means we are relying on uptime from the hub for tests to pass

tiborvass commented:
@jfrazelle @dmcgowan i'm not that okay with relying on the hub for tests.

If we nevertheless want to do that, then we should not have the tests fail if the hub fails. Or maybe we should condition the test based on a first ping and if the ping works but not the pull then fail.

tianon commented:
I'm definitely with @tiborvass here -- we've removed tests in the past
specifically because they relied on the uptime of the Hub, and that bit us
often enough and painfully enough that we had to claw it back.

Assuming you've got a recent build of the "Dockerfile", the test suite
ought to be able to run completely internet-less, IMO.

jfrazelle commented:
ah something for a 502 maybe? (if you want because I don't want to hold
this up and inconvience the registry team who has worked so hard this week)
I would be willing to send a secondary patch for that

On Thu, Feb 5, 2015 at 3:06 PM, Tibor Vass <notifications@github.com> wrote:

> @jfrazelle <https://github.com/jfrazelle> @dmcgowan
> <https://github.com/dmcgowan> i'm not that okay with relying on the hub
> for tests.
>
> If we nevertheless want to do that, then we should not have the tests fail
> if the hub fails. Or maybe we should condition the test based on a first
> ping and if the ping works but not the pull then fail.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10594#issuecomment-73149530>.
>

tiborvass commented:
@jfrazelle I'd much rather have their registry in the dockerfile and do whatever they would do on the hub but locally.

And totally agree that we can carry the PR, they've been working very hard. Just wanted to voice my opinion :)

jfrazelle commented:
yes true, I was actually the one who changed even the robots.txt to not
rely on the hub (ah the horror) ok I am reconsidering my LGTM now what do
you think no test? or we change this test to make sure the verification
message is not there because it is not verified (is that cheating)

On Thu, Feb 5, 2015 at 3:10 PM, Tianon Gravi <notifications@github.com>
wrote:

> I'm definitely with @tiborvass here -- we've removed tests in the past
> specifically because they relied on the uptime of the Hub, and that bit us
> often enough and painfully enough that we had to claw it back.
>
> Assuming you've got a recent build of the "Dockerfile", the test suite
> ought to be able to run completely internet-less, IMO.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/docker/pull/10594#issuecomment-73150041>.
>

tiborvass commented:
@jfrazelle either we find a way of using the v2 registry in the dockerfile, or we cheat for now and use the hub without making the tests fail for it. We can rewrite the test later once we have a better idea for registry tests.

dmcgowan commented:
My choices were completely remove the test or pull from the hub as is done in 2 other tests already.  I know everyone likes red, but not when its just tests.  Either way is fine with me, but if you are going to reject that part of this PR for pulling from the hub, then the 2 other tests which also do so should be updated as well  (TestPullImageFromCentralRegistry in docker_cli_pull_test.go and TestEventsImagePull in docker_cli_events_test.go).

I am also +1 for not relying on the hub, just in this case would take significantly more changes to get local verification to work.

tiborvass commented:
@dmcgowan Thanks for bringing those tests to our attention!

These are the options i'm willing to consider:
* make an exception coz @dmcgowan is cool
* remove the test from this PR and add it back correctly in another PR fixing the two other tests as well
* make this test not fail if the ping to the hub fails and fix the two other tests in another PR

dmcgowan commented:
I will explore having the test skip if the error is network related

dmcgowan committed:
Update verification message logic

Only show the verification message if all the tarsum checks pass and the image manifest is verified.
No longer return an error when a tarsum verification fails, just reset the verification flag.
Tarsum verification is less meaningful without a verified manifest and therefore it should not cause an error.
Updated the verified image test to pull an image which expected to have a verified manifest and contents.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

dmcgowan commented:
Pushed an update which skips the test when an error occurs on pull.  Unfortunately I was not able to isolate the network error condition which does make this test less useful.  However it will still catch bad changes to verification flow.

tiborvass commented:
LGTM

jfrazelle commented:
LGTM

jfrazelle commented:
ping @icecrime i wil let you merge bc you are release cheif hehe

icecrime commented:
LGTM (keeping this in a corner of my head if a 1.5.0-rc5 happens, but I don't think it alone mandates for one).
